Valigator.Core/Data.cs

```diff
@@ -117,5 +119,70 @@ private static IDataContainer<TValue> GetErrorDataContainer(IDataContainer<TValu
 
 		public static implicit operator TValue(Data<TValue> data)
 			=> data.Value;
+
+		public override bool Equals(object obj) 
+			=> obj is Data<TValue> data 
+				&& this == data;
+
+		public static bool operator ==(Data<TValue> x, Data<TValue> y)
+			=> x.DataDescriptor.Equals(y.DataDescriptor)
+					&& (x._value.Equals(y._value)
+						|| (IsCollection(x) && IsCollection(y) && IsSequenceEqual(x._value as ICollection, y._value as ICollection)));
+
+
+
+		private static bool IsSequenceEqual(ICollection a, ICollection b)
+		{
+			if (a.Count == b.Count)
+			{
+				var firstEnumerator = a.GetEnumerator();
+				var secondEnumerator = b.GetEnumerator();
+				for (int i = 0; i < a.Count; i++)
+				{
+					firstEnumerator.MoveNext();
+					secondEnumerator.MoveNext();
+					if (!(IsCollection(firstEnumerator.Current) && IsCollection(secondEnumerator.Current) ?
+						IsSequenceEqual(firstEnumerator.Current as ICollection, secondEnumerator.Current as ICollection)
+						: AreObjectsEqual(firstEnumerator.Current, secondEnumerator.Current)))
+						return false;
+				}
+				return true;
+			}
+			return false;		
+		}
+
+		private static bool AreObjectsEqual(object a, object b)
+		{
+			if (a.GetType() == b.GetType())
+				return a.Equals(b);
+			else
+				return false;
+		}
+		private static bool IsCollection(object a) 
+			=> a.GetType().GetInterface(nameof(ICollection)) != null;
+
+		private static bool IsCollection<T>(Data<T> a)
+			=> typeof(T).GetInterface(nameof(ICollection)) != null;
+
+		public static bool operator !=(Data<TValue> x, Data<TValue> y)
+			=> !x._value.Equals(y._value);
+
+		public override int GetHashCode()
+		{
+			int hashCode = 943777100;
+			if (IsCollection(this))
```

Comment:
I think this would probably be a little cleaner if you passed in TValue as either a generic or a parameter directly. I would advise something like I suggested above.